### DISCUSSION TITLE
Flutter FCM for iOS - Reports Issue

### DISCUSSION THREAD

**Discussion by karthikeyanp307** [firebase/flutterfire#5705] [Q&A]
Hi,

I am using firebase_messaging  v9.0.1 in my Flutter application. On configuring this library based on the https://pub.dev/packages/firebase_messaging/example, I am able to receive the notification for android in both foreground and background states. But the same for iOS is not working. I am getting the below error,

APNS device token not set before retrieving FCM Token for Sender ID ''. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.

Do I want to call any other function apart from FirebaseMessaging.instance.getToken() for iOS? Please help.

**Comment by darshankawar** [comment]
@karthikeyanp307 
Can you check and confirm if your iOS device is connected properly to the internet ?
If so, does it return token from `FirebaseMessaging.instance.getToken()` ?

Also, provide your `flutter doctor -v` to properly address this issue.
Check if any of these helps.

https://stackoverflow.com/a/66123064

**Reply by georgehoss** [reply]
> Hi,
> 
> I am using firebase_messaging v9.0.1 in my Flutter application. On configuring this library based on the https://pub.dev/packages/firebase_messaging/example, I am able to receive the notification for android in both foreground and background states. But the same for iOS is not working. I am getting the below error,
> 
> APNS device token not set before retrieving FCM Token for Sender ID ''. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
> 
> Do I want to call any other function apart from FirebaseMessaging.instance.getToken() for iOS? Please help.

I had the same problem, I went step by step again with firebase documentation and realize that the push notification capability was missing on debug and release modes. 

Steps to solve the issue:

1. Go to xcode
2. Select runner 
3. Signing & Capabilities 
4. Check debug and release capabilities you need to have background modes and push notification

[![push notification capability][1]][1]

https://firebase.flutter.dev/docs/messaging/apple-integration


  


  [1]: https://i.stack.imgur.com/gbwwp.png

**Reply by rak54** [reply]
> @karthikeyanp307 Can you check and confirm if your iOS device is connected properly to the internet ? If so, does it return token from `FirebaseMessaging.instance.getToken()` ?
> 
> Also, provide your `flutter doctor -v` to properly address this issue. Check if any of these helps.
> 
> https://stackoverflow.com/a/66123064

EveryThing is fine by side . Though I have deleted APNS (.p8) key that i am using and created new one . Could that be the problem ?

**Comment by sguru049** [comment]
@darshankawar I'm also facing the same issue for my web-app, with firebase_messaging v9.0.1. The notifications doesn't work in iOS devices, for any browsers. For Android, It works good for Chrome but fails for Firefox and Opera.

`FirebaseMessaging.instance.getToken()` returns `null` for all browsers on iOS. Same for Firefox and Opera on Android. The issue is, the notification asking for permission to send notifications doesn't pop-up on these devices. 

`flutter doctor -v` also shows "No issues Found!"

Can you please suggest a solution ?

**Comment by peteroleeman** [comment]
having same issue as well

**Comment by karthikeyanp307** [comment]
@darshankawar  - I tried with the https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example. When configured with FCM - google-services it works fine for android and getting the same error as above for iOS. 

"APNS device token not set before retrieving FCM Token for Sender ID ''. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set." My iOS device is connected to internet and there were no network related issues while running this.

Also getAPNSToken() always returns null. Please help me on this. 

Thanks.

**Comment by darshankawar** [comment]
@karthikeyanp307 
It seems you haven't setup your iOS part properly. See this document and follow the steps as directed, https://firebase.flutter.dev/docs/messaging/apple-integration.

Also, the document states that, `On iOS & macOS, you must have correctly setup your project to integrate with FCM and APNs.`

**Reply by lulamprea** [reply]
Any update on the fixing of this bug?

**Comment by karthikeyanp307** [comment]
@darshankawar - Thanks for your reply. I have followed the steps provided in the reference link.

I used firebase_messaging: ^7.0.3 earlier with the same application and both android & iOS were working fine. When upgraded the same application to the latest firebase messaging v9.0.1 I started getting this error for iOS.But the same works fine for android. Is there anything specific that we need to do, while upgrading to the latest firebase messaging library. 

Thanks.

**Reply by georgehoss** [reply]
I had the same problem, I went step by step again with firebase documentation and realize that the push notification capability was missing on debug and release modes. 

Steps to solve the issue:

1. Go to xcode
2. Select runner 
3. Signing & Capabilities 
4. Check debug and release capabilities you need to have background modes and push notification

[![push notification capability][1]][1]

https://firebase.flutter.dev/docs/messaging/apple-integration


  


  [1]: https://i.stack.imgur.com/gbwwp.png

**Reply by rak54** [reply]
> I had the same problem, I went step by step again with firebase documentation and realize that the push notification capability was missing on debug and release modes.
> 
> Steps to solve the issue:
> 
> 1. Go to xcode
> 2. Select runner
> 3. Signing & Capabilities
> 4. Check debug and release capabilities you need to have background modes and push notification
> 
> [![push notification capability](https://camo.githubusercontent.com/03e3976f29a856a4d9668ebbe45664715c0ce7edaaa0e4f3bfcc133599d571f8/68747470733a2f2f692e737461636b2e696d6775722e636f6d2f67627777702e706e67)](https://i.stack.imgur.com/gbwwp.png)
> 
> https://firebase.flutter.dev/docs/messaging/apple-integration

I have done all that , still not working . Anything else ?

**Comment by karthikeyanp307** [comment]
@darshankawar  - I have re-verified the steps in the reference link,  but facing the same issue still now.
Please let me know if you need any more details to help me on this. 

Thanks.

**Comment by lulamprea** [comment]
@darshankawar, I have been following this thread for a while due to I'm facing the same issue, is this something that google is interested to fix? 

Is there any information that you require so we can help to make progress in this bug? . Perhaps if you can point us in to the firebase module source code we could try to contribute fixing the code?

Thank you very much for your help

**Comment by darshankawar** [comment]
Cc: @Salakar for further insights.

**Reply by lulamprea** [reply]
Thank you

**Reply by lulamprea** [reply]
@Salakar is there anything that we can do to help here?

**Comment by sudhanshug16** [comment]
Facing the same issue with v9.1.2. Trying to make it work with a new app and getting the following error:
```
APNS device token not set before retrieving FCM Token for Sender ID '<>'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
```
`.getAPNSToken()` returns `null`

I have followed all the steps mentioned in documentation, checked it twice but still doesn't work. Android works fine.

**Comment by karthikeyanp307** [comment]
@Salakar - Could you please help me on resolving this. Is there anything missing in the documentation that I am supposed to do to make this work? 

Let me know for any clarification or inputs to help us on this.

**Comment by themaaz32** [comment]
I am also getting the same error with null APNs Token

**Comment by Salakar** [comment]
For messaging related issues it's helpful to provide a lot more detail on your problem to allow maintainers such as myself to help on these issues, e.g.;

 - How are you sending the notification? (if relevant)
 - What's your full notification payload including FCM options?  (if relevant)
 - Full code showing how you've registered your background handler, not just copy paste the functions, we need to see explicitly where you've called it and where your handler is located, e.g. it should be a top level function.  (if relevant)
 - What device are you testing on, os version, make, model.
 - What specifically isn't working, background, foreground, seeing logs but not seeing notifications, etc, details here help.  (if relevant)
 - Versions, Flutter, FlutterFire messaging, etc.
 - Related screenshots & log outputs.

**Comment by karthikeyanp307** [comment]
Hi @Salakar ,

Thanks for your reply. We are trying to assign custom notification sound while sending notification to iOS device. But we are not even receiving the notification when tried with the latest firebase_messaging library.

Below are the required details.

•	We are sending the notification via Postman for testing and in production via Spring boot application.

o	URL - https://fcm.googleapis.com/fcm/send

o	Body
```
{
"to": "token",
"notification": {
"sound": "alert.caf",
"title": "message_title",
"body": "message_body",
"id": "message_id"
},
"data": {
"body": "message_body",
"category": "category",
"title": "message_title",
"click_action": "FLUTTER_NOTIFICATION_CLICK",
"id": "message_id"
},
"apns": {
"headers": {
"apns-priority": "10"
},
"payload": {
"aps": {
"sound": "alert.caf"
}
}
}
}
```

•	Source Code -  https://github.com/karthikeyanp307/my_flutter_fcm_app/tree/master/my_app (We are using your example project and have setup the fcm google service info into it). I have removed the google services file for android and iOS app.

•	Device Info - iPhone 8; OS v13.5.1; Apple

•	Notification are being sent but not received in device in all the app states i.e., foreground, background and when app is killed

•	Version Info
    o	Flutter 2.0.1 • channel stable • https://github.com/flutter/flutter.git
    o	Framework • revision c5a4b4029c (8 weeks ago) • 2021-03-04 09:47:48 -0800
    o	Engine • revision 40441def69
    o	Tools • Dart 2.12.0
    
    o	firebase_core: "1.0.2"
    o	firebase_messaging: "9.1.0"

•	Logs

```
Launching lib/main.dart on iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: XXXXXX
Running Xcode build...
Xcode build done.                                           56.0s
Installing and launching...
Debug service listening on ws://127.0.0.1:56262/94AfOt510kA=/ws
Syncing files to device iPhone...
7.3.0 - [Firebase/Messaging][I-FCM002022] APNS device token not set before retrieving FCM Token for Sender ID '934113297926'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
flutter: FCM Token: dBNhstxWtE3somntJQ92R8:APA91bHGbOvsnSN54PqN9oV3RTmB3fd8cCpwtkjuOphudD2Q7P_mzj57yF_P_seH5raSKD0iATIUMxMq6y8QY9jB_HrV7gzHwYd7QLWLR9y-RkZIMACK_Kfu_rF_VPjPS_JcVeX9UHpH
```

Kindly help us in resolving this. Thanks in advance. Let us know if you need any more details regarding.

**Comment by karthikeyanp307** [comment]
Hi @Salakar ,

Adding to the above post. 

I have used the FCM Sample application downloaded from https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example and just added the google services json for iOS.

Let us know if you need any more details.

**Comment by Anas-Othman** [comment]
i am facing the same issue . my app was running with fcm 7.0.3 and everything was fine for android and ios , but when upgrading to 9.1.1 got the APNS issue  , and FirebaseMessaging.instance.getToken() always returns null too. 

i haven't try to send a notification to a physical ios device yet ,  is that error means the notification will not be sent ?

**Reply by Anas-Othman** [reply]
APNS device token not set before retrieving FCM Token for Sender ID ''. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
this error appears only when run the app on the ios simulator but when running on a physical everything is working well and the notifications arrived on foreground and background

**Reply by karthikeyanp307** [reply]
Hi Anas,

Thanks for sharing this. But I am running the app in real device but still facing this issue. Have you did anything apart from those given in the example app and FCM documentation. If so share us with those.

Thanks.

**Reply by dizdaroglu** [reply]
Hi,  Is the problem solved? I am getting the same error and I don't know the solution.

**Reply by wepashka** [reply]
Any solution?

**Comment by Anas-Othman** [comment]
Hi , No i didn't do anything different ... i just used only the getToken()
as usual and everything worked fine on the real device

**Reply by dizdaroglu** [reply]
Thank you i have to try it on real device right away 👍

**Comment by dizdaroglu** [comment]
hi, it didn't work although i tested it on real device. I have followed all the steps mentioned in documentation, but doesn't work.

**Comment by lulamprea** [comment]
Is there any progress in this matter?  I have not been able to resolve this case, so  I'm developing a native  IOS plugin. It seems that is the solution while the Flutter team fix the issue.

**Reply by dizdaroglu** [reply]
I couldn't figure it out, I'm still trying

**Comment by sudhanshug16** [comment]
For me the fix was to add push notifications as a capability. I mean I added it already but under the wrong subtab. This is how I added the capability 
![image](https://user-images.githubusercontent.com/9914480/117663549-65843d80-b1be-11eb-8ac2-c1b61e85e861.png)

I thought "All", "Debug", etc are mere filters 🤦🏼  but turns out that I had to select "All" and then add the capability. This is more of a UX bug than anything.

**Comment by shinewanna** [comment]
Facing same issue, it work well before update, after update iOS firebase_messaging is showing this warning and not receiving notifications any more.

APNS device token not set before retrieving FCM Token for Sender ID ''. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.

**Comment by hosein-nadali** [comment]
Same issue

**Comment by luvishq** [comment]
Anyone able to resolve this issue ? @Salakar @karthikeyanp307 ?

**Comment by pkgit-tech** [comment]
same issue any update

**Comment by icodeyou** [comment]
`flutter clean` solved this issue for me

**Comment by dhruv048** [comment]
Hey, even i am getting same issue
Works like a charm in android, but i am getting this issue on IOS. 

> APNS device token not set before retrieving FCM Token for Sender ID '458258671137'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.

when trying to retrieve APNS token through 
`FirebaseMessaging.instance.getAPNSToken().then((String token) {
        print(token);
      });`
it returns null.
i have followed all steps shown in [https://firebase.flutter.dev/docs/messaging/apple-integration](url).
Tried everything suggested in this discussion but still the issue presist.
Thanks,

**Comment by himsa** [comment]
same issue, any update ?

**Comment by irufano** [comment]
same isssue here any update ??

**Comment by RawahaOsi** [comment]
The audacity to close, lock and move such a thread from issues to discussion is out of this world. FlutterFire dudes should be paying more attention to such blocker issues that are breaking thousands of apps worldwide.

**Reply by papirovdavid** [reply]
Same issue for me: on Real IOS device,  ver 7.x notifications worked, but in 9.x - NOT. 
Need your urgently help!

**Reply by hdn-luan** [reply]
Same problem herer :/

**Comment by Nitingadhiya** [comment]
@papirovdavid please upgrade with latest version, issue resolved.

**Reply by hdn-luan** [reply]
Still with the same problem :/

**Comment by danderer** [comment]
Same issue.  Preceded by "The default Firebase app has not yet been configured."  The doc for doing a CLI initialization makes no mention of the GoogleServices-Info.plist file for iOS, and so I didn't install it.

Installing GoogleServices-Info.plist allowed FlutterFire to successfully initialize, and for messaging via APNS to work for me.

**Reply by garyhaywood** [reply]
I did the same thing - started with a new project 3 times and each time the same error.
Installed GoogleServices-Info.plist and used await Firebase.initializeApp();

The project fired up working straight away.

**Reply by c0c4i** [reply]
Wow! I did the same thing, dropping GoogleServices-Info.plist inside project folder from Xcode and work.

Can someone explain why this things it's not mentioned on FlutterFire Documentation?

**Reply by t-hubb** [reply]
Solved the issue for me as well, thanks!

**Reply by wepashka** [reply]
Same issue any update?

**Reply by wepashka** [reply]
It didn't help at all, any other solution please?

**Reply by wepashka** [reply]
It was a low internet connection on my phone internet

**Comment by Riya-Singhal15** [comment]
I've tried all the solutions, but on some devices, it's giving this error. Please provide a solution if anyone has resolved it. Thanks.
